tests/catch/unit/memory/hipFree.cc

```diff
@@ -0,0 +1,557 @@
+#include "DriverContext.hh"
+
+/*
+ * This testcase verifies [ hipFree || hipFreeArray || hipArrayDestroy || hipHostFree with
+ * hipHostMalloc ]
+ * 1. Check that hipFree implicitly synchronises the device.
+ * 2.Perform multiple allocations and then call hipFree on each pointer concurrently (from unique
+ * threads) for different memory types and different allocation sizes.
+ * 3. Pass nullptr as argument and check that no operation is performed and hipSuccess is returned.
+ * 4. Pass an invalid ptr and check that hipErrorInvalidValue is returned.
+ * 5. Call hipFree twice on the same pointer and check that the implementation handles the second
+ * call correctly.
+ * 6. HipHostFree only:
+ *    Try to free memory that has been registered with hipHostRegister and check that
+ * hipErrorInvalidValue is returned.
+ */
+
+
+enum FreeType { DevFree, ArrayFree, ArrayDestroy, HostFree };
+
+__global__ void waitKernel(clock_t offset) {
+  auto time = clock();
+  while (clock() - time < offset) {
+  }
+}
+
+static __global__ void clock_kernel(clock_t clock_count, size_t* clockOut, size_t* co = nullptr) {
+  *clockOut = clock();
+  clock_t start_clock = clock();
+  clock_t clock_offset = 0;
+  while (clock_offset < clock_count) {
+    clock_offset = clock() - start_clock;
+  }
+  if (co != nullptr) *co = clock_offset;
+}
+
+// number of clocks the device is running at (device frequency)
+static size_t ticks = 0;
+
+// helper function used to set the device frequency variable
+
+static size_t findTicks() {
+  hipDeviceProp_t prop;
+  int device;
+  size_t *clockFromKernel, *clockOffset;
+  HIP_CHECK(hipMalloc(&clockFromKernel, sizeof(size_t)));
+  HIP_CHECK(hipMalloc(&clockOffset, sizeof(size_t)));
+  HIP_CHECK(hipGetDevice(&device));
+  HIP_CHECK(hipGetDeviceProperties(&prop, device));
+
+  constexpr float mseconds = 1000;
+  constexpr float error = 0.02 * mseconds;
+
+  clock_t devFreq = static_cast<clock_t>(prop.clockRate);  // in kHz
+  clock_t time = devFreq * mseconds;
+
+
+  while (1) {
+    auto start = std::chrono::high_resolution_clock::now();
+    hipLaunchKernelGGL(clock_kernel, dim3(1), dim3(1), 0, 0, time, clockFromKernel, clockOffset);
+    HIP_CHECK(hipGetLastError());
+    HIP_CHECK(hipDeviceSynchronize());
+    auto stop = std::chrono::high_resolution_clock::now();
+    auto result = std::chrono::duration<double, std::milli>(stop - start).count();
+    size_t co = 0;
+    HIP_CHECK(hipMemcpy(&co, clockOffset, sizeof(size_t), hipMemcpyDeviceToHost));
+    if (result >= (mseconds - error) && result <= (mseconds + error)) {
+      HIP_CHECK(hipFree(clockFromKernel));
+      HIP_CHECK(hipFree(clockOffset));
+      return co;
+
+    } else {
+      auto off = fabs(mseconds - result) / mseconds;
+      if (result >= mseconds) {
+        time -= (time * off);
+      } else {
+        time += (time * off);
+      }
+    }
+  }
+}
+
+static void runKernelForMs(size_t ms, hipStream_t stream = nullptr) {
+  if (ticks == 0) {
```

Comment:
This all assumes you will always be using the same device and only a single thread (setting ticks could occur on multiple devices at once). This may be fine for this test, but might get a little dodgy if you try to use this in more places. 
If its fine for now then I suppose an appropriate response is just to detail these restrictions on the function.

Reply:
Added comments explaining lack of thread safety, as well as approach to take if changing devices.